### PR TITLE
Fix issues that came up with building cuDF with main

### PR DESCRIPTION
**Pull request by miscco**
We need `cuda_runtime.h`for the C++ `cudaMallocHost` and not `cuda_runtime_api.h`

### CODE REVIEW

File: cub/cub/util_device.cuh [hunk 1]
```diff
@@ -49,14 +49,12 @@
 
 #include <cub/detail/device_synchronize.cuh>
 #include <cub/util_debug.cuh>
-#include <cub/util_ptx.cuh>
 #include <cub/util_type.cuh>
+#include <cub/util_vsmem.cuh>
 // for backward compatibility
 #include <cub/util_temporary_storage.cuh>
 
-#include <cuda/discard_memory>
```
Comment (miscco) [inline cub/cub/util_device.cuh:57]: @gevtushenko can you please give this refactoring a look?

File: thrust/thrust/system/cuda/detail/copy_if.h [hunk 2]
```diff
@@ -181,7 +180,7 @@ struct DispatchCopyIf
     CUDA_CUB_RET_IF_FAIL(status);
     OffsetT num_selected = get_value(policy, d_num_selected_out);
 
-    output_end = output + num_selected;
+    output += num_selected;
```
Comment (miscco) [inline thrust/thrust/system/cuda/detail/copy_if.h:183]: Should this be `thrust::advance`?

File: thrust/thrust/system/cuda/detail/copy_if.h [hunk 3]
```diff
@@ -147,7 +146,7 @@ struct DispatchCopyIf
     // Return for empty problems
     if (num_items == 0)
     {
-      output_end = output;
+      output;
```
Comment (bernhardmgruber) [inline thrust/thrust/system/cuda/detail/copy_if.h:149]: Statement without side effects. Can be removed.

File: cub/cub/util_device.cuh (hunk 1 above)
Comment (jrhemstad) [reply cub/cub/util_device.cuh:57]: What's the story with the changes to `util_device.cuh`?

File: thrust/thrust/system/cuda/detail/copy_if.h (hunk 3 above)
Comment (miscco) [reply thrust/thrust/system/cuda/detail/copy_if.h:149]: 🤦 obviously it was already too late

File: cub/cub/util_device.cuh [hunk 4]
```diff
@@ -49,14 +49,12 @@
...
 #include <cuda/std/type_traits>
-#include <cuda/std/utility>
```
Comment (gevtushenko) [inline cub/cub/util_device.cuh:59]: important: `::cuda::std::forward` is used below. Let's return this include.

File: thrust/thrust/system/cuda/detail/copy_if.h (hunk 2 above)
Comment (gevtushenko) [reply thrust/thrust/system/cuda/detail/copy_if.h:183]: yes

**Review by gevtushenko — CHANGES_REQUESTED** [review]

File: cub/cub/util_vsmem.cuh [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+   *
+   * @note Needs to be followed by `__syncthreads()` if the function returns true and the virtual shared memory is
+   * supposed to be reused after this function call.
+   */
+  template <bool needs_vsmem_ = needs_vsmem, typename ::cuda::std::enable_if<needs_vsmem_, int>::type = 0>
+  static _CCCL_DEVICE _CCCL_FORCEINLINE bool discard_temp_storage(typename AgentT::TempStorage& temp_storage)
+  {
+#  if defined(_CCCL_CUDA_COMPILER)
+    // Ensure all threads finished using temporary storage
+    NV_IF_TARGET(NV_IS_HOST, (__syncthreads();))
```
[the remark this answers is not in the document]
Comment (miscco) [reply cub/cub/util_vsmem.cuh:191]: Oh yeah That should go back to CTA_SYNC();

File: cub/cub/util_vsmem.cuh [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+   * passed to the agent, specialized for the case when we can use native shared memory as temporary
+   * storage and taking a linear block id.
+   */
+  static _CCCL_DEVICE __forceinline__ typename AgentT::TempStorage&
+  get_temp_storage(typename AgentT::TempStorage& static_temp_storage, vsmem_t&, std::size_t)
+  {
+    return static_temp_storage;
+  }
+
+#  if defined(_CCCL_CUDA_COMPILER)
```
[the remark this answers is not in the document]
Comment (miscco) [reply cub/cub/util_vsmem.cuh:140]: The issue is that some rapids configurations build with only a host compiler and that should continue to work.

It was broken by us pulling in `cuda/ptx` and also using cuda specific code in the vsmem-helper

File: cub/cub/util_device.cuh [hunk 7]
```diff
@@ -49,14 +49,12 @@
 
 #include <cub/detail/device_synchronize.cuh>
 #include <cub/util_debug.cuh>
-#include <cub/util_ptx.cuh>
 #include <cub/util_type.cuh>
+#include <cub/util_vsmem.cuh>
```
[the remark this answers is not in the document]
Comment (miscco) [reply cub/cub/util_device.cuh:53]: I checked and it turned out to be not so few places

It lead to a lot of "incomplete type" errors in the PreviousPolicy template argument

File: cub/cub/util_vsmem.cuh [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+  // The amount of shared memory or virtual shared memory required by the algorithm's agent
+  static constexpr std::size_t required_smem = sizeof(typename AgentT::TempStorage);
+
+  // Whether we need to allocate global memory-backed virtual shared memory
+  static constexpr bool needs_vsmem
+#  if defined(_CCCL_CUDA_COMPILER)
+    = required_smem > max_smem_per_block;
+#  else // ^^^ _CCCL_CUDA_COMPILER ^^^ / vvv !_CCCL_CUDA_COMPILER vvv
+    = false;
+#  endif // !_CCCL_CUDA_COMPILER
```
[the remark this answers is not in the document]
Comment (miscco) [reply cub/cub/util_vsmem.cuh:105]: Same as above there is existing code that relies on some header including this to be usable in a non-CUDA context.

Breaking this on a minor version is icky, even if it was not supported

File: cub/cub/util_device.cuh (hunk 7 above)
[the remark this answers is not in the document]
Comment (gevtushenko) [reply cub/cub/util_device.cuh:53]: Do you have a standalone reproducer?

File: thrust/thrust/optional.h [hunk 9]
```diff
@@ -838,7 +838,7 @@ struct nullopt_t
 /// void foo (thrust::optional<int>);
 /// foo(thrust::nullopt); //pass an empty optional
 /// ```
-static constexpr nullopt_t nullopt{nullopt_t::do_not_use{}, nullopt_t::do_not_use{}};
+_LIBCUDACXX_CPO_ACCESSIBILITY nullopt_t nullopt{nullopt_t::do_not_use{}, nullopt_t::do_not_use{}};
```
Comment (miscco) [inline thrust/thrust/optional.h:841]: We should double check visibility issues in old standard modes

File: cub/cub/util_device.cuh (hunk 1 above)
Comment (miscco) [reply cub/cub/util_device.cuh:57]: Just including it in the right places works

File: cub/cub/util_device.cuh (hunk 7 above)
[the remark this answers is not in the document]
Comment (miscco) [reply cub/cub/util_device.cuh:53]: It turned out to be a snafu with me pulling out the policy wrappers. Fixed

File: cub/cub/util_vsmem.cuh (hunk 8 above)
[the remark this answers is not in the document]
Comment (miscco) [reply cub/cub/util_vsmem.cuh:108]: Fixed

File: cub/cub/util_vsmem.cuh (hunk 6 above)
[the remark this answers is not in the document]
Comment (miscco) [reply cub/cub/util_vsmem.cuh:143]: Fixed

File: thrust/thrust/optional.h (hunk 9 above)
Comment (miscco) [reply thrust/thrust/optional.h:841]: Trying out keeping the static in all standard modes

File: libcudacxx/include/cuda/std/__utility/pair.h [hunk 10]
```diff
@@ -283,18 +283,21 @@ struct _LIBCUDACXX_TEMPLATE_VIS pair : public __pair_base<_T1, _T2>
...
 
-  template <class _U1          = _T1,
-            class _U2          = _T2,
-            class _Constraints = typename __pair_constraints<_T1, _T2>::template __constructible<_U1, _U2>,
+  template <class _U1, class _U2>
+  using __pair_constructible = typename __pair_constraints<_T1, _T2>::template __constructible<_U1, _U2>;
+
+  template <class _U1                                                  = _T1,
+            class _U2                                                  = _T2,
+            class _Constraints                                         = __pair_constructible<_U1, _U2>,
```
Comment (griwes) [inline libcudacxx/include/cuda/std/__utility/pair.h:291]: :/

File: thrust/thrust/tuple.h [hunk 11]
```diff
@@ -94,28 +94,32 @@ _CCCL_HOST_DEVICE inline bool operator>(const null_type&, const null_type&)
...
+{};
 
 /*! \brief \p tuple is a class template that can be instantiated with up to ten
- *  arguments. Each template argument specifies the type of element in the \p
+ *  otheruments. Each template otherument specifies the type of element in the \p
  *  tuple. Consequently, tuples are heterogeneous, fixed-size collections of
- *  values. An instantiation of \p tuple with two arguments is similar to an
- *  instantiation of \p pair with the same two arguments. Individual elements
+ *  values. An instantiation of \p tuple with two otheruments is similar to an
+ *  instantiation of \p tuple with the same two otheruments. Individual elements
```
Comment (griwes) [inline thrust/thrust/tuple.h:122]: Otheruments? :P

**Review by griwes — APPROVED** [review]

File: thrust/thrust/tuple.h (hunk 11 above)
Comment (miscco) [reply thrust/thrust/tuple.h:122]: Oh man search and replace hits

**Comment by miscco** [comment]
download

File: libcudacxx/include/cuda/functional [hunk 12]
```diff
@@ -68,6 +68,8 @@
 #include <cuda/std/type_traits>
 #include <cuda/std/functional>
 #include <cuda/std/utility>
+#include <cuda/std/__algorithm/min.h>
+#include <cuda/std/__algorithm/max.h>
```
Comment (gevtushenko) [inline libcudacxx/include/cuda/functional:72]: question: why are these needed in functional?

File: thrust/thrust/tuple.h [hunk 13]
```diff
@@ -94,28 +95,32 @@ _CCCL_HOST_DEVICE inline bool operator>(const null_type&, const null_type&)
  *  \tparam N This parameter selects the element of interest.
  *  \tparam T A \c tuple type of interest.
  *
- *  \see pair
+ *  \see tuple
```
Comment (gevtushenko) [inline thrust/thrust/tuple.h:98]: suggestion: this should stay pair

File: thrust/thrust/tuple.h [hunk 14]
```diff
@@ -94,28 +95,32 @@ _CCCL_HOST_DEVICE inline bool operator>(const null_type&, const null_type&)
...
+template <class>
+struct __is_tuple_of_iterator_references : _CUDA_VSTD::false_type
+{};
 
 /*! \brief \p tuple is a class template that can be instantiated with up to ten
  *  arguments. Each template argument specifies the type of element in the \p
  *  tuple. Consequently, tuples are heterogeneous, fixed-size collections of
  *  values. An instantiation of \p tuple with two arguments is similar to an
- *  instantiation of \p pair with the same two arguments. Individual elements
+ *  instantiation of \p tuple with the same two arguments. Individual elements
```
Comment (gevtushenko) [inline thrust/thrust/tuple.h:123]: suggestion: it should stay as `\p pair`

File: thrust/thrust/tuple.h [hunk 15]
```diff
@@ -143,19 +148,89 @@ using tuple_size = ::cuda::std::tuple_size<T>;
  *  }
  *  \endcode
  *
- *  \see pair
+ *  \see tuple
```
Comment (gevtushenko) [inline thrust/thrust/tuple.h:151]: suggestion: this should stay `pair`

File: libcudacxx/include/cuda/functional (hunk 12 above)
Comment (miscco) [reply libcudacxx/include/cuda/functional:72]: This is because of backwards compatibility. We had them in `functional` and we do not expose `__algorithm` yet

File: thrust/thrust/tuple.h (hunk 13 above)
Comment (miscco) [reply thrust/thrust/tuple.h:98]:
```suggestion
 *  \see pair
```

File: thrust/thrust/tuple.h (hunk 14 above)
Comment (miscco) [reply thrust/thrust/tuple.h:123]:
```suggestion
 *  instantiation of \p pair with the same two arguments. Individual elements
```

File: thrust/thrust/tuple.h (hunk 15 above)
Comment (miscco) [reply thrust/thrust/tuple.h:151]:
```suggestion
 *  \see pair
```

**Review by gevtushenko — APPROVED** [review]

File: cub/cub/util_vsmem.cuh [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+    return true;
+  }
+};
+
+template <class DefaultAgentT, class FallbackAgentT>
+constexpr bool use_fallback_agent()
+{
+  return (sizeof(typename DefaultAgentT::TempStorage) > max_smem_per_block)
+      && (sizeof(typename FallbackAgentT::TempStorage) <= max_smem_per_block);
+}
```
Comment (bernhardmgruber) [inline cub/cub/util_vsmem.cuh:195]: This must compile with C++11, right? Otherwise: an inline variable may be a good fit here.

File: cub/cub/util_vsmem.cuh [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+/**
+ * @brief Class template that helps to prevent exceeding the available shared memory per thread block.
+ *
+ * @tparam AgentT The agent for which we check whether per-thread block shared memory is sufficient or whether virtual
+ * shared memory is needed.
+ */
+template <typename AgentT>
+class vsmem_helper_impl
+{
+private:
```
Comment (bernhardmgruber) [inline cub/cub/util_vsmem.cuh:82]: Matter of style:
```suggestion
class vsmem_helper_impl
{
```

File: cub/cub/util_vsmem.cuh [hunk 18]
```diff
@@ -0,0 +1,255 @@
...
+ * requirements of the default agent exceed `max_smem_per_block`, yet the shared memory requirements of the fallback
+ * agent falls within the bounds of `max_smem_per_block`.
+ * @tparam FallbackAgentT The fallback agent, instantiated with the given fallback tuning policy
+ */
+template <typename DefaultAgentPolicyT,
+          typename DefaultAgentT,
+          typename FallbackAgentPolicyT = DefaultAgentPolicyT,
+          typename FallbackAgentT       = DefaultAgentT,
+          bool UseFallbackPolicy        = use_fallback_agent<DefaultAgentT, FallbackAgentT>()>
+struct vsmem_helper_with_fallback_impl : public vsmem_helper_impl<DefaultAgentT>
```
Comment (bernhardmgruber) [inline cub/cub/util_vsmem.cuh:218]: Matter of style:
```suggestion
struct vsmem_helper_with_fallback_impl : vsmem_helper_impl<DefaultAgentT>
```

File: cub/cub/util_vsmem.cuh [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+  using agent_t        = DefaultAgentT;
+  using agent_policy_t = DefaultAgentPolicyT;
+};
+template <typename DefaultAgentPolicyT, typename DefaultAgentT, typename FallbackAgentPolicyT, typename FallbackAgentT>
+struct vsmem_helper_with_fallback_impl<DefaultAgentPolicyT, DefaultAgentT, FallbackAgentPolicyT, FallbackAgentT, true>
+    : public vsmem_helper_impl<FallbackAgentT>
+{
+  using agent_t        = FallbackAgentT;
+  using agent_policy_t = FallbackAgentPolicyT;
+};
```
Comment (bernhardmgruber) [inline cub/cub/util_vsmem.cuh:229]: Suggestion: maybe use two conditionals instead of specialization to shorten code:
```suggestion
template <typename DefaultAgentPolicyT,
          typename DefaultAgentT,
          typename FallbackAgentPolicyT = DefaultAgentPolicyT,
          typename FallbackAgentT       = DefaultAgentT>
struct vsmem_helper_with_fallback_impl : public vsmem_helper_impl<DefaultAgentT>
{
  static constexpr bool useFallbackPolicy = use_fallback_agent<DefaultAgentT, FallbackAgentT>()
  using agent_t        = std::conditional_t<useFallbackPolicy, FallbackAgentT, DefaultAgentT>;
  using agent_policy_t = std::conditional_t<useFallbackPolicy, FallbackAgentPolicyT, DefaultAgentPolicyT>;
};
```

File: cub/cub/util_vsmem.cuh [hunk 20]
```diff
@@ -0,0 +1,255 @@
...
+                                  FallbackPolicyT,
+                                  AgentT<FallbackPolicyT, AgentParamsT...>>;
+
+/**
+ * @brief Alias template for the `vsmem_helper_t` by using a simple fallback policy that uses `DefaultPolicyT` as basis,
+ * overwriting `64` threads per block and `1` item per thread.
+ */
+template <typename DefaultPolicyT, template <typename...> class AgentT, typename... AgentParamsT>
+using vsmem_helper_default_fallback_policy_t =
+  vsmem_helper_fallback_policy_t<DefaultPolicyT, policy_wrapper_t<DefaultPolicyT, 64, 1>, AgentT, AgentParamsT...>;
```
Comment (bernhardmgruber) [inline cub/cub/util_vsmem.cuh:249]: The documentation is lying to me. It says the alias is to `vsmem_helper_t`, but it aliases to `vsmem_helper_fallback_policy_t`. This may be a copy&paste error from above?

File: thrust/testing/pair.cu [hunk 21]
```diff
@@ -322,4 +322,16 @@ void TestPairStructuredBindings(void)
   ASSERT_EQUAL(b, b2);
 }
 DECLARE_UNITTEST(TestPairStructuredBindings);
-#endif
+
+void TestPairCTAD(void)
```
Comment (bernhardmgruber) [inline thrust/testing/pair.cu:326]: Question: any reason to keep the `void` other than (very) old habits? We are not aiming to compile with C?
```suggestion
void TestPairCTAD()
```
Second occurance later.

File: thrust/thrust/tuple.h [hunk 22]
```diff
@@ -98,18 +99,22 @@ _CCCL_HOST_DEVICE inline bool operator>(const null_type&, const null_type&)
...
+using tuple_element = _CUDA_VSTD::tuple_element<N, T>;
 
 /*! This metafunction returns the number of elements
  *  of a \p tuple type of interest.
  *
  *  \tparam T A \c tuple type of interest.
  *
- *  \see pair
+ *  \see tuple
  *  \see tuple
```
Comment (bdice) [inline thrust/thrust/tuple.h:110]: This is duplicated.
```suggestion
 *  \see tuple
```

**Review by bdice — APPROVED** [review]
I don't know the full context for all of these changes but they look fine from what I know. Thanks!

File: thrust/thrust/pair.h [hunk 23]
```diff
@@ -73,10 +73,49 @@ using tuple_size = ::cuda::std::tuple_size<T>;
...
+#endif // _CCCL_STD_VER >= 2017
+
+template <class T1, class T2>
+inline _CCCL_HOST_DEVICE
+  _CUDA_VSTD::__enable_if_t<_CUDA_VSTD::__is_swappable<T1>::value && _CUDA_VSTD::__is_swappable<T2>::value, void>
+  swap(pair<T1, T2>& lhs, pair<T1, T2>& rhs) noexcept(
+    (_CUDA_VSTD::__is_nothrow_swappable<T1>::value && _CUDA_VSTD::__is_nothrow_swappable<T2>::value))
+{
+  lhs.swap(rhs);
+}
```
Comment (bernhardmgruber) [inline thrust/thrust/pair.h:107]: Question: why isn't `swap` a hidden friend?

File: thrust/thrust/pair.h [hunk 24]
```diff
@@ -73,10 +73,49 @@ using tuple_size = ::cuda::std::tuple_size<T>;
...
+  _CCCL_HOST_DEVICE constexpr pair(_U1&& __u1, _U2&& __u2)
+      : super_t(_CUDA_VSTD::forward<_U1>(__u1), _CUDA_VSTD::forward<_U2>(__u2))
+  {}
+#endif // _CCCL_COMPILER_GCC < 9 || _CCCL_COMPILER_CLANG < 12
+};
+
+#if _CCCL_STD_VER >= 2017
+template <class _T1, class _T2>
+_CCCL_HOST_DEVICE pair(_T1, _T2) -> pair<_T1, _T2>;
+#endif // _CCCL_STD_VER >= 2017
```
Comment (bernhardmgruber) [inline thrust/thrust/pair.h:98]: Question: doesn't this fail for compilers with partial C++17 support and `-std=c++1z`?
Suggestion: test for the feature instead:
```suggestion
#ifdef __cpp_deduction_guides
template <class _T1, class _T2>
_CCCL_HOST_DEVICE pair(_T1, _T2) -> pair<_T1, _T2>;
#endif // __cpp_deduction_guides
```

File: thrust/thrust/pair.h [hunk 25]
```diff
@@ -88,3 +127,38 @@ using ::cuda::std::make_pair;
...
+namespace std
+{
+template <class T1, class T2>
+struct tuple_size<THRUST_NS_QUALIFIER::pair<T1, T2>> : tuple_size<pair<T1, T2>>
+{};
+
+template <size_t Id, class T1, class T2>
+struct tuple_element<Id, THRUST_NS_QUALIFIER::pair<T1, T2>> : tuple_element<Id, pair<T1, T2>>
+{};
+} // namespace std
```
Comment (bernhardmgruber) [inline thrust/thrust/pair.h:163]: Matter of style, but saves a few LoCs:
```suggestion
template <class T1, class T2>
struct ::std::tuple_size<THRUST_NS_QUALIFIER::pair<T1, T2>> : std::tuple_size<std::pair<T1, T2>>
{};

template <size_t Id, class T1, class T2>
struct ::std::tuple_element<Id, THRUST_NS_QUALIFIER::pair<T1, T2>> : std::tuple_element<Id, std::pair<T1, T2>>
{};
```

File: thrust/thrust/tuple.h (hunk 22 above)
Comment (bernhardmgruber) [inline thrust/thrust/tuple.h:110]: Question: unintended replace? Docs linked to `pair` and `tuple` before and now link to `tuple` twice.

File: thrust/thrust/tuple.h [hunk 26]
```diff
@@ -150,12 +155,82 @@ using tuple_size = ::cuda::std::tuple_size<T>;
...
+
+#if defined(_CCCL_COMPILER_MSVC_2017)
+  // MSVC2017 needs some help to convert tuples
+  template <class... Us,
+            _CUDA_VSTD::__enable_if_t<!_CUDA_VSTD::is_same<tuple<Us...>, tuple>::value, int> = 0,
+            _CUDA_VSTD::__enable_if_t<_CUDA_VSTD::__tuple_convertible<_CUDA_VSTD::tuple<Us...>, super_t>::value, int> = 0>
+  _CCCL_HOST_DEVICE constexpr operator tuple<Us...>()
+  {
+    return __to_tuple<Us...>(typename _CUDA_VSTD::__make_tuple_indices<sizeof...(Ts)>::type{});
+  }
```
Comment (bernhardmgruber) [inline thrust/thrust/tuple.h:191]: Question: does this function need a `const` overload, or is it just an MSVC workaround that doesn't require a conversion from a const tuple?

File: thrust/thrust/tuple.h [hunk 27]
```diff
@@ -150,12 +155,82 @@ using tuple_size = ::cuda::std::tuple_size<T>;
...
+
+  template <class... Us, size_t... Id>
+  _CCCL_HOST_DEVICE constexpr tuple<Us...> __to_tuple(_CUDA_VSTD::__tuple_indices<Id...>) const
+  {
+    return tuple<Us...>{_CUDA_VSTD::get<Id>(*this)...};
+  }
+#endif // _CCCL_COMPILER_MSVC_2017
+};
+
+#if _CCCL_STD_VER >= 2017
```
Comment (bernhardmgruber) [inline thrust/thrust/tuple.h:201]: Again, maybe use a feature test instead.

File: thrust/thrust/tuple.h [hunk 28]
```diff
@@ -150,12 +155,82 @@ using tuple_size = ::cuda::std::tuple_size<T>;
...
+#endif // _CCCL_STD_VER >= 2017
+
+template <class... Ts>
+inline _CCCL_HOST_DEVICE
+  _CUDA_VSTD::__enable_if_t<_CUDA_VSTD::__all<_CUDA_VSTD::__is_swappable<Ts>::value...>::value, void>
+  swap(tuple<Ts...>& __x,
+       tuple<Ts...>& __y) noexcept((_CUDA_VSTD::__all<_CUDA_VSTD::__is_nothrow_swappable<Ts>::value...>::value))
+{
+  __x.swap(__y);
+}
```
Comment (bernhardmgruber) [inline thrust/thrust/tuple.h:219]: Again: why not a hidden friend?

File: cub/cub/util_vsmem.cuh (hunk 17 above)
Comment (miscco) [reply cub/cub/util_vsmem.cuh:82]: While this is indeed superfluous the goal is to better delineate the public and the internal interface

File: thrust/thrust/pair.h (hunk 25 above)
Comment (bernhardmgruber) [reply thrust/thrust/pair.h:163]: Alright, got your point! We also have `tuple_size`, etc. in `::cuda::std`, so it is not automatically clear which one we are specializing.

File: thrust/thrust/pair.h (hunk 23 above)
Comment (bernhardmgruber) [reply thrust/thrust/pair.h:107]: Ah, I get it. Because:
```c++
template <typename T1, typename T2>
struct pair {
  friend _CUDA_VSTD::__enable_if_t<_CUDA_VSTD::__is_swappable<T1>::value && 
            _CUDA_VSTD::__is_swappable<T2>::value, void>
   swap(pair<T1, T2>& lhs, pair<T1, T2>& rhs) noexcept { ... }
};
```
would immediately fail for a `pair<nonswappable, int>`. You want to delay that failure to whether `swap` is actually called. So you would need:
```c++
template <typename T1, typename T2>
struct pair {
  template <typename U1 = T1, typename U2 = T2>
  friend _CUDA_VSTD::__enable_if_t<_CUDA_VSTD::__is_swappable<U1>::value && 
            _CUDA_VSTD::__is_swappable<U2>::value, void> ...
};
```
And I agree, this may be worse than where we started. Thx!

File: cub/cub/util_vsmem.cuh (hunk 19 above)
Comment (bernhardmgruber) [reply cub/cub/util_vsmem.cuh:229]: > I do not believe this is equivalent, Once we have vsmem_helper_impl<DefaultAgentT> and once vsmem_helper_impl<FallbackAgentT>

Right, that was an error on my side.

> Also std::conditional adds another two more instantiations for relatively litte gain as the classes are so simple

Aren't the types already instantiated, since they are template parameters? I am not forming any new templates.

**Review by bernhardmgruber — APPROVED** [review]

File: cub/cub/util_vsmem.cuh (hunk 19 above)
Comment (miscco) [reply cub/cub/util_vsmem.cuh:229]: But you are instantiating `cuda::std::conditional`

File: thrust/thrust/pair.h [hunk 29]
```diff
@@ -73,10 +73,49 @@ using tuple_size = ::cuda::std::tuple_size<T>;
  *          provided by <tt>pair::second_type</tt>.
  */
 template <class T, class U>
-using pair = ::cuda::std::pair<T, U>;
-
```
Comment (bernhardmgruber) [inline thrust/thrust/pair.h:77]: I just fell over this again because of a doxygen issue and now wonder, why didn't you just pull `pair` into the `thrust` namespace without creating an alias? Like:
```c++
using _CUDA_VSTD::pair;
```
This works with CTAD and passes the unit tests.
